src/Disks/DiskRemote.cpp

```diff
@@ -0,0 +1,442 @@
+#include "DiskRemote.h"
+#include <Common/createHardLink.h>
+#include "DiskFactory.h"
+
+#include <Disks/IO/ReadBufferFromRemoteDisk.h>
+#include <Disks/IO/WriteBufferFromRemoteDisk.h>
+#include <Disks/LocalDirectorySyncGuard.h>
+#include <Disks/ObjectStorages/DiskObjectStorage.h>
+#include <Interpreters/Context.h>
+#include <Common/StringUtils/StringUtils.h>
+#include <Common/atomicRename.h>
+#include <Common/filesystemHelpers.h>
+#include <Common/quoteString.h>
+
+#include <fstream>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+#include <Disks/DiskFactory.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <Common/logger_useful.h>
+#include <Common/randomSeed.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+std::mutex DiskRemote::reservation_mutex;
+
+using DiskRemotePtr = std::shared_ptr<DiskRemote>;
+
+class DiskRemoteReservation : public IReservation
+{
+public:
+    DiskRemoteReservation(const DiskRemotePtr & disk_, UInt64 size_, UInt64 unreserved_space_)
+        : disk(disk_), size(size_), unreserved_space(unreserved_space_)
+    {
+    }
+
+    UInt64 getSize() const override { return size; }
+    UInt64 getUnreservedSpace() const override { return unreserved_space; }
+
+    DiskPtr getDisk(size_t i) const override
+    {
+        if (i != 0)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Can't use i != 0 with single disk reservation. It's a bug");
+        return disk;
+    }
+
+    Disks getDisks() const override { return {disk}; }
+
+    void update(UInt64 new_size) override
+    {
+        std::lock_guard lock(DiskRemote::reservation_mutex);
+        disk->reserved_bytes -= size;
```

Comment:
do we really need to implement reservation on the client size? seems like it should also check with remote server?